apisix/plugins/limit-count/delayed_syncer.lua

```diff
@@ -0,0 +1,132 @@
+-- specific language governing permissions and limitations
+-- under the License.
+
+local core = require("apisix.core")
+local consts = require("apisix.constants")
+local new_redis_cli = require("apisix.utils.redis").new
+local new_redis_cluster_cli = require("apisix.utils.rediscluster").new
+
+local ngx_timer = ngx.timer
+local redis_stop
+local redis_cluster_stop
+local _M = {}
+
+
+local function sync_counter_data(premature, counter, to_be_synced, redis_confs, script)
+  if premature then
+    return
+  end
+  for key, _ in pairs(to_be_synced) do
+    local num_reqs, err = counter:get(key .. consts.REDIS_COUNTER)
+    if not num_reqs then
+      core.log.error("failed to get num_reqs shdict during periodic sync: ", err)
+      return
+    end
+
+    local conf = redis_confs[key]
+
+    local red, err
+    if conf.policy == "redis" then
+      red, err = new_redis_cli(conf)
+    elseif conf.policy == "redis-cluster" then
+      red, err = new_redis_cluster_cli(conf)
+    else
+      core.log.error("invalid policy type: ", conf.policy)
+      return
+    end
+
+    if not red then
+      core.log.error("failed to get redis client during periodic sync: ", err)
+      return
+    end
+
+    local res, err = red:eval(script, 1, key, conf.count, conf.time_window, num_reqs)
+    if err then
+      core.log.error("failed to sync shdict data to redis: ", err)
+      return
+    end
+
+    local remaining = res[1]
+    local ttl = res[2]
+    core.log.info("syncing shdict num_req counter to redis. remaining: ", remaining,
+                  " ttl: ", ttl, " reqs: ", num_reqs)
+    counter:set(key .. consts.SHDICT_REDIS_REMAINING, tonumber(remaining), tonumber(ttl))
+    counter:set(key .. consts.REDIS_COUNTER, 0)
+
+    if (not redis_stop and (conf.policy == "redis"))
+        or (not redis_cluster_stop and (conf.policy == "redis-cluster")) then
+      local ok, err = ngx_timer.at(conf.sync_interval, sync_counter_data, counter, to_be_synced,
+                                   redis_confs, script)
+      if not ok then
+        core.log.error("failed to create redis syncer timer: ", err,
+                        ". New main redis syncer will be created.")
+
+        -- next incoming request will pick this up and create a new timer
+        counter:set(consts.REDIS_SYNCER, false)
+      end
+    end
+  end
+end
+
+function _M.rate_limit_with_delayed_sync(conf, counter, to_be_synced, redis_confs, key, cost,
+                                         limit, window, script)
+  local syncer_started = counter:get(consts.REDIS_SYNCER)
+  if not syncer_started then
+    local ok, err = ngx_timer.at(conf.sync_interval, sync_counter_data, counter, to_be_synced,
+                                 redis_confs, script)
+    if ok then
+      counter:set(consts.REDIS_SYNCER, true)
+    else
+      core.log.error("failed to create main redis syncer timer: ", err, ". Will retry next time.")
+    end
+  end
+
+  to_be_synced[key] = true -- add to table for syncing
```

Comment:
`to_be_synced & redis_confs` are module scoped variables, worker processes that not running `syncer` would not able to tell `syncer` to sync the keys added by them through this way

Reply:
For the `one syncer per worker process` scenario, it's OK.

BTW:
1. `to_be_synced & redis_confs` are not necessary pass to ngx.timer.at as extra args?
2.  using redis client instance instead of redis config? 
     could be like this:  `to_be_synced[key] = { redis_cli = red_cli }`?

Reply:
> using redis client instance instead of redis config?

cosockets cannot be shared to ngx.timer functions.

Reply:
> to_be_synced & redis_confs are module scoped variables

I get it now. We need to make `to_be_synced` and `redis_confs` visible to the worker process which has `syncer`.

Thanks for your review again.

Reply:
> > to_be_synced & redis_confs are module scoped variables
> 
> I get it now. We need to make `to_be_synced` and `redis_confs` visible to the worker process which has `syncer`.
> 
> Thanks for your review again.

:D

Yes.
But these module variables are originally visible to the `per worker process syncer`.

Reply:
> > using redis client instance instead of redis config?
> 
> cosockets cannot be shared to ngx.timer functions.

May be we could consider other implementation instead of ngx.timer. e.g. `request driven syncer`:
if reach` count or time threshold`, sync. immediately, in-place, in the req. handler context